servicetalk-http-api/src/main/java/io/servicetalk/http/api/OffloadingFilter.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.api;
+
+import io.servicetalk.concurrent.api.Executor;
+import io.servicetalk.concurrent.api.Single;
+
+import static java.util.function.Function.identity;
+
+/**
+ * An {@link StreamingHttpServiceFilterFactory} implementation which offloads filters using a provided strategy.
+ */
+final class OffloadingFilter implements StreamingHttpServiceFilterFactory, HttpExecutionStrategyInfluencer {
+
+    private final HttpExecutionStrategy strategy;
+    private final StreamingHttpServiceFilterFactory offloaded;
+
+    /**
+     * @param strategy Execution strategy for the offloaded filters
+     * @param offloaded Filters to be offloaded
+     */
+    OffloadingFilter(HttpExecutionStrategy strategy, StreamingHttpServiceFilterFactory offloaded) {
+        this.strategy = strategy;
+        this.offloaded = offloaded;
+    }
+
+    @Override
+    public StreamingHttpServiceFilter create(StreamingHttpService service) {
+        return new StreamingHttpServiceFilter(offloaded.create(service)) {
+
+            @Override
+            public Single<StreamingHttpResponse> handle(final HttpServiceContext ctx,
+                                                        StreamingHttpRequest request,
+                                                        final StreamingHttpResponseFactory responseFactory) {
+                Executor se = strategy.executor();
+                Executor e = null != se ? se : ctx.executionContext().executor();
+
+                // The service should see our ExecutionStrategy inside the ExecutionContext:
+                final HttpServiceContext wrappedCtx =
+                        new ExecutionContextOverridingServiceContext(ctx, strategy, e);
+
+                if (strategy.isDataReceiveOffloaded()) {
+                    request = request.transformMessageBody(p -> p.publishOn(e));
```

Comment:
IIRC, with the 0.41 behavior of `publishOn` operator, the move of offloading logic from `NettyHttpServer` to this filter won't change the fact that offloading is applied at the beginning of this publisher chain. If this is the case, let's keep the code here to simplify `NettyHttpServer`, but hide the `appendNonOffloadingServiceFilter` methods on `HttpServerBuilder` to prevent users from using them. We can clarify in the PR title/message that this change doesn't add a user-facing feature.